impeller/renderer/render_pass.cc

```diff
@@ -98,4 +98,71 @@ const std::weak_ptr<const Context>& RenderPass::GetContext() const {
   return context_;
 }
 
+void RenderPass::SetPipeline(
+    const std::shared_ptr<Pipeline<PipelineDescriptor>>& pipeline) {
+  pending_.pipeline = pipeline;
+}
+
+void RenderPass::SetCommandLabel(const std::string& label) {
+#ifdef IMPELLER_DEBUG
+  pending_.label = label;
+#endif  // IMPELLER_DEBUG
+}
+
+void RenderPass::SetStencilReference(uint32_t value) {
+  pending_.stencil_reference = value;
+}
+
+void RenderPass::SetBaseVertex(uint64_t value) {
+  pending_.base_vertex = value;
+}
+
+void RenderPass::SetViewport(Viewport viewport) {
+  pending_.viewport = viewport;
+}
+
+void RenderPass::SetScissor(IRect scissor) {
+  pending_.scissor = scissor;
+}
+
+void RenderPass::SetInstanceCount(size_t count) {
+  pending_.instance_count = count;
+}
+
+bool RenderPass::SetVertexBuffer(VertexBuffer buffer) {
+  return pending_.BindVertices(std::move(buffer));
+}
+
+bool RenderPass::Dispatch() {
+  auto result = AddCommand(std::move(pending_));
+  pending_ = Command{};
+  return result;
+}
```

Comment:
There are probably opportunities to cut down API traffic by not having draw calls reset all state, but starting with this seems like a good idea.

Reply:
Ah in fact I think you mentioned this earlier w.r.t. the viewport and scissor.

Reply:
Yeah, this change is mostly pretty dumb as its keeping the exact same implementation.

Reply:
My mind keeps wandering to this patch tonight, so figured I'd jot down my thonks in more detail. ;) (No action necessary)

I think the winning long game here is to have no implicit draw state management behavior in `RenderPass` at all (except for the default state, of course).

It would be great if we end up in a place where we don't implicitly reset _any_ state when recording a draw call, since that would maximally open the door for us to optimize traffic within high context parts of the pass workload (for example, any `Contents` that performs multiple draws).

Of course, we'd still probably want utilities like `RenderPass::ResetState()` for low context transitions (like going from one `Contents` to another).
The refactor in Entities for this would be ezpz. All we'd need to get started is something like:
* Refactor `Contents` to use a stub pattern -- non-virtual `Contents::Render()` calls virtual `Contents::OnRender()` -- and have `Contents::Render()` call `render_pass->ResetState()` (or maybe have an Entities-specific utility for this that ignores stuff like viewport/scissor because we know we're not using them right now) just before dispatching to `Contents::OnRender()`.
* Then, for each `Contents::OnRender()` impl we have that records multiple draws, don't reset stuff in cases where it's obviously unnecessary. For example, if the draws are using the same pipeline, you probably don't need to reset anything at all.

Reply:
I agree with not resetting the state. Really, that is just working against ourselves since we also have the pass binding ccache essentially undoing that work at the render pass level. I suspect we will still need some state tracking for bindings, but perhaps we can get sufficiently clever as to make that irrelevant.